rmw_zenoh_cpp/src/detail/rmw_data_types.cpp

```diff
@@ -471,25 +466,34 @@ void sub_data_handler(
       "rmw_zenoh_cpp", "Unable to obtain sequence number from the attachment.");
   }
 
-  int64_t source_timestamp = get_int64_from_attachment(&sample->attachment, "source_timestamp");
+  int64_t source_timestamp = get_int64_from_attachment(attachment, "source_timestamp");
   if (source_timestamp < 0) {
     // We failed to get the source timestamp from the attachment.  While this
     // isn't fatal, it is unusual and so we should report it.
     source_timestamp = 0;
     RMW_ZENOH_LOG_ERROR_NAMED(
-      "rmw_zenoh_cpp", "Unable to obtain sequence number from the attachment.");
+      "rmw_zenoh_cpp",
+      "Unable to obtain sequence number from the attachment.");
   }
 
+  const z_loaned_bytes_t * payload = z_sample_payload(sample);
+
+  z_owned_slice_t slice;
+  z_bytes_deserialize_into_slice(payload, &slice);
+
   sub_data->add_new_message(
     std::make_unique<saved_msg_data>(
-      zc_sample_payload_rcinc(sample),
-      sample->timestamp.time, pub_gid, sequence_number, source_timestamp), z_loan(keystr));
+      slice,
```

Comment:
I don't think keeping an owned type as we do in other parts is an issue. `z_move` will turn an owned type into a moved type. I think it would change a lot in RMW.